### PR TITLE
#1360: proposal: remove runtime prop types checking (closes #1360)

### LINKED ISSUE #1360: proposal: remove runtime prop types checking
With all of our project now using TypeScript, and TS being quite adopted in the React community in general, we could remove prop types via tcomb `props` in order to:

1. spare some code and a runtime dependency
2. avoid cases where the two typings (runtime and static) differ in a subtle way

This should be done on a few other separate repos I think, such as `react-flexview`

### PR DESCRIPTION
**Pull request by giogonzo**
Closes #1360

As expected, the only interesting cases are the one previously involving tcomb `refinement`s. In a single case, it was possible to encode it at type-level (`TabbedPanel` tabs should be non-empty). For all the other cases (TimePicker, Meter mainly) I didn't invest time in expressing the invariants at type level, only reused the same checks performed at runtime inside of `componentDidMount`, and warning in case of issues.

## Test Plan

All checks are passing. This PR should have no impact at runtime, and I decided to not mark it as breaking for this reason. Let me know your thoughts

### tests performed

> A Test Plan is used to show what you tested to make sure your code works fine. You should write here all that you did to test, and provide some results of your testing.

> These results can be screenshots, query results, or even just pointers to unit tests that successfully passed.

### tests not performed (domain coverage)

> At times not everything can be tested, and writing what hasn't been tested is just as important as writing what has been tested.

> An example of partial test is a field displaying 4 possible values. If 3 values are tested, with screenshots, and 1 is not, then it should be mentioned here.

### CODE REVIEW

File: src/Button/Button.tsx [hunk 1]
```diff
@@ -170,6 +126,14 @@ export class Button extends React.PureComponent<Button.Props> {
     style: {}
   };
 
+  componentDidMount() {
+    if (process.env.NODE_ENV !== "production") {
+      if (!satisfyAll(...propsInvariants)(this.props)) {
```
Comment (veej) [inline src/Button/Button.tsx]: is this spread useful?

File: src/FocusableView/FocusableView.tsx [hunk 2]
```diff
@@ -133,7 +121,7 @@ export class FocusableView extends React.Component<FocusableView.Props> {
     return React.createElement(
       component as any,
       locals,
-      t.Function.is(children) ? children(focused) : children
+      isFunction(children) ? children(focused) : children
```
Comment (veej) [inline src/FocusableView/FocusableView.tsx]: why not `typeof children === "function"` as you did for strings and numbers?
That's the same exact thing lodash's `isFunction` does.

File: src/TimePicker/TimePicker.tsx [hunk 3]
```diff
@@ -58,28 +53,19 @@ const getComponents = (
...
-    (!value || (lteTime(value, maxTime) && lteTime(minTime, value))),
-  "Props"
-);
+function validProps({ value, minTime, maxTime }: TimePicker.Props): boolean {
+  const validTimes =
+    (value == null || isTime(value)) &&
+    (minTime == null || isTime(minTime)) &&
+    (maxTime == null || isTime(maxTime));
+  const minMaxCoherent =
+    minTime != null && maxTime != null && lteTime(minTime, maxTime);
```
Comment (veej) [inline src/TimePicker/TimePicker.tsx]: Shouldn't this be `minTime == null || maxTime == null || lteTime(minTime, maxTime)`?

File: src/Button/Button.tsx (hunk 1 above)
Comment (giogonzo) [reply src/Button/Button.tsx]: not sure, I left it as I found it in the code previously

File: src/FocusableView/FocusableView.tsx (hunk 2 above)
Comment (giogonzo) [reply src/FocusableView/FocusableView.tsx]: makes sense, I'll do it

File: src/TimePicker/TimePicker.tsx (hunk 3 above)
Comment (giogonzo) [reply src/TimePicker/TimePicker.tsx]: yup, good catch, thanks

File: src/Scroll/ScrollView.tsx [hunk 4]
```diff
@@ -223,7 +226,7 @@ export class ScrollView extends React.Component<ScrollView.Props> {
           this.scrollView = sv;
         }}
       >
-        {t.Function.is(children) ? children(this.scrollTo) : children}
+        {isFunction(children) ? children(this.scrollTo) : children}
```
Comment (veej) [inline src/Scroll/ScrollView.tsx]: replace with `typeof children === "function"`?

File: src/Tablo/Cell/Cell.tsx [hunk 5]
```diff
@@ -101,15 +85,11 @@ export function _Cell<T extends {}, K extends keyof T | never = never>(
           vAlignContent={vAlignContent}
           hAlignContent={hAlignContent}
         >
-          {t.Function.is(render) ? render(data, rowData, rowIndex) : render}
+          {isFunction(render) ? render(data, rowData, rowIndex) : render}
```
Comment (veej) [inline src/Tablo/Cell/Cell.tsx]: replace with `typeof render === "function"`?

File: src/Scroll/ScrollView.tsx (hunk 4 above)
Comment (giogonzo) [reply src/Scroll/ScrollView.tsx]: This change fails with a weird error:
```ts
This expression is not callable.
  Not all constituents of type '((scrollTo: ScrollTo) => Children) | (string & {}) | (number & {}) | (false & {}) | (true & {}) | (((scrollTo: ScrollTo) => Children) & string) | (((scrollTo: ScrollTo) => Children) & number) | ... 4 more ... | (((scrollTo: ScrollTo) => Children) & ReactPortal)' are callable.
    Type 'string & {}' has no call signatures.ts(2349)
```

Not sure how to fix without using an `any` cast really..

**Pull request merged**